### PR TITLE
Initial integration for Twitter

### PR DESCRIPTION
**Pull request by bwillard**
This supports exporting images (that aren't retweets) from twitter.

It also supports importing an image stream into Twitter.  As a use case this seems a little sketchy as it would spam your profile, but I wanted it mostly as an example.

TODO: support periscope video transfer.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -18,6 +18,10 @@ matrix:
     # language: java causes Travis to automatically run Gradle tests based on the
     # presence of the appropriate build files.
     - language: java # Gradle (used automatically)
+      install:
```
Comment (rtannenbaum) [inline .travis.yml]: can you rebase against master? i think this is from an earlier commit

Comment (bwillard) [reply .travis.yml]: Done.

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterApiWrapper.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+final class TwitterApiWrapper {
+  static  Twitter getInstance(
+      AppCredentials appCredentials,
+      TokenSecretAuthData authData) {
+
+      ConfigurationBuilder cb = new ConfigurationBuilder();
+      cb.setDebugEnabled(false)
+          .setOAuthConsumerKey(appCredentials.getKey())
+          .setOAuthConsumerSecret(appCredentials.getSecret())
+          // TODO: I think the token/secret expire, we need to check into refreshing them
```
Comment (holachuy) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterApiWrapper.java]: For refresh, there's no interface or class but for Google see data-transfer-project/extensions/data-transfer/portability-data-transfer-google/src/main/java/org/dataportabilityproject/datatransfer/google/common/GoogleCredentialFactory.java

File: extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthDataGenerator.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    return new AuthFlowConfiguration(
+        requestToken.getAuthorizationURL(),
+        new TokenSecretAuthData(requestToken.getToken(), requestToken.getTokenSecret()));
+  }
+
+  @Override
+  public AuthData generateAuthData(
+      String callbackBaseUrl, String authCode, String id, AuthData initialAuthData, String extra) {
+
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthDataGenerator.java]: nit: extra line

File: extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthServiceExtension.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.spi.api.auth.AuthDataGenerator;
+import org.dataportabilityproject.spi.api.auth.AuthServiceProviderRegistry.AuthMode;
+import org.dataportabilityproject.spi.api.auth.extension.AuthServiceExtension;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TwitterAuthServiceExtension implements AuthServiceExtension {
```
Comment (rtannenbaum) [inline extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthServiceExtension.java]: nit: documentation since public

File: extensions/cloud/portability-cloud-local/build.gradle [hunk 5]
```diff
@@ -17,4 +17,8 @@
 dependencies {
     compile project(':portability-spi-cloud')
     compile project(':portability-spi-transfer')
+
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-local/build.gradle]: nit: consider moving this to a separate PR since it's not related to the twitter integration

File: extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/LocalJobStore.java [hunk 6]
```diff
@@ -26,12 +27,15 @@
 import java.util.Map.Entry;
 import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/LocalJobStore.java]: move this to a separate PR too

also, I thought the logger should come from this extension (the cloud-local extension)'s context (ExtensionContext.getLogger) but it seems we're not doing that elsewhere. MIght want to check with @jimmarino on that but I'm also fine with leaving this alone as it's consistent with how it's currently done elsewhere

File: extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/LocalJobStore.java [hunk 7]
```diff
@@ -123,9 +129,12 @@ public synchronized UUID findFirst(JobAuthorization.State jobState) {
     // Mimic an index lookup
     for (Entry<UUID, Map<String, Object>> job : JOB_MAP.entrySet()) {
       Map<String, Object> properties = job.getValue();
-      if (JobAuthorization.State.valueOf(
-              properties.get(PortabilityJob.AUTHORIZATION_STATE).toString())
-          == jobState) {
+      State state = State.valueOf(
+          properties.get(PortabilityJob.AUTHORIZATION_STATE).toString());
+
+      logger.debug("checking job {} in state {} to see if it matches {}",
```
Comment (rtannenbaum) [inline extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/LocalJobStore.java]: nit: message could be clarified, how about:

looking up first job in state {}: found job {} (state {})

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosExporter.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import twitter4j.MediaEntity;
+import twitter4j.Paging;
+import twitter4j.ResponseList;
+import twitter4j.Status;
+import twitter4j.Twitter;
+import twitter4j.TwitterException;
+
+public class TwitterPhotosExporter implements
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosExporter.java]: nit: documentation

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosExporter.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      }
+    }
+    Paging paging = new Paging(pageNumber, PAGE_SIZE);
+    try {
+      logger.debug("Getting tweets for {} (page {})", twitterApi.getId(), pageNumber);
+      ResponseList<Status> statuses = twitterApi.getUserTimeline(twitterApi.getId(), paging);
+      List<PhotoModel> photos = new ArrayList<>();
+      for (Status status : statuses) {
+        boolean hasMedia = status.getMediaEntities().length > 0;
+        if (hasMedia & !status.isRetweet()) {
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosExporter.java]: && ?

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+import org.dataportabilityproject.types.transfer.auth.TokenSecretAuthData;
+import org.dataportabilityproject.types.transfer.models.photos.PhotoModel;
+import org.dataportabilityproject.types.transfer.models.photos.PhotosContainerResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import twitter4j.StatusUpdate;
+import twitter4j.Twitter;
+import twitter4j.TwitterException;
+
+public class TwitterPhotosImporter
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java]: documentation

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+  public TwitterPhotosImporter(AppCredentials appCredentials) {
+    this.appCredentials = appCredentials;
+  }
+
+  @Override
+  public ImportResult importItem(UUID jobId, TokenSecretAuthData authData,
+      PhotosContainerResource data) {
+    Twitter twitterApi = TwitterApiWrapper.getInstance(appCredentials, authData);
+    // Don't know what to do with albums :(.
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java]: TODO / github issue?

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        twitterApi.tweets().updateStatus(update);
+      } catch (IOException | TwitterException e) {
+        logger.warn("problem importing twitter photo", e);
+        return new ImportResult(ResultType.ERROR, e.getMessage());
+      }
+    }
+    return new ImportResult(ResultType.OK);
+  }
+
+  private InputStream getImageAsStream(String urlStr) throws IOException {
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java]: I think Anna moved this into a library, might have been google-specific. perhaps it could be moved into libraries/ ?

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterTransferExtension.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
+import org.dataportabilityproject.spi.transfer.extension.TransferExtension;
+import org.dataportabilityproject.spi.transfer.provider.Exporter;
+import org.dataportabilityproject.spi.transfer.provider.Importer;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TwitterTransferExtension implements TransferExtension {
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterTransferExtension.java]: documentation

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterTransferExtension.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  private static final List<String> SUPPORTED_TYPES = ImmutableList.of("photos");
+  private static final Logger LOGGER = LoggerFactory.getLogger(TwitterTransferExtension.class);
+  private static final String TWITTER_KEY = "TWITTER_KEY";
```
Comment (rtannenbaum) [inline extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterTransferExtension.java]: could these be defined in a central place since they are also defined in the auth extension? Not sure if we do that for other extensions, so I'm fine leaving it alone too

File: extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthDataGenerator.java (hunk 3 above)
Comment (bwillard) [reply extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthDataGenerator.java]: Done.

File: extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthServiceExtension.java (hunk 4 above)
Comment (bwillard) [reply extensions/auth/portability-auth-twitter/src/main/java/org/dataportabilityproject/auth/twitter/TwitterAuthServiceExtension.java]: Done

File: extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/LocalJobStore.java (hunk 7 above)
Comment (bwillard) [reply extensions/cloud/portability-cloud-local/src/main/java/org/dataportabilityproject/cloud/local/LocalJobStore.java]: Done.

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosExporter.java (hunk 9 above)
Comment (bwillard) [reply extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosExporter.java]: Done.

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java (hunk 10 above)
Comment (bwillard) [reply extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java]: Made private

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java (hunk 11 above)
Comment (bwillard) [reply extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java]: I flushed out the comment, I don't know that there is anything to do with this expect note it.

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java (hunk 12 above)
Comment (bwillard) [reply extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterPhotosImporter.java]: Ya I'll take a look at that in a separate PR.

File: extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterTransferExtension.java (hunk 14 above)
Comment (bwillard) [reply extensions/data-transfer/portability-data-transfer-twitter/src/main/java/org/dataportabilityproject.transfer.twitter/TwitterTransferExtension.java]: I think we should move these key values out of the plugins in general and have the framework only send the right values  on to the plugins.  Right now Twitter gets to see all the other plugin's vlaues, which isn't ideal.
